### ISSUE TITLE
`cbn` and `/=` have different behaviors when `SimplIsCbn`

### ISSUE THREAD

**Issue by pi8027** [coq/coq#17251]
#### Description of the problem

```coq
From Coq Require Import ssreflect.

Set SimplIsCbn.

Section Iteration.
Variable T : Type.

Fixpoint iter (n : nat) (f : T -> T) (x : T) :=
  match n with S i => f (iter i f x) | 0 => x end.

Lemma iterSr n f x : iter (S n) f x = iter n f (f x).
Proof.
elim: n => //.
move=> /=. (* no effect *)
cbn. (* some occurrences of [iter] are reduced. *)
Abort.

End Iteration.
```

#### Coq Version

8.16.1

**Comment by pi8027** [comment]
Apparently `move=> /=`, `cbn`, and `rewrite /=` have three distinct behaviors:
```coq
From Coq Require Import ssreflect.

Set SimplIsCbn.

Section Iteration.
Variable T : Type.

Definition iter (n : nat) (f : T -> T) (x : T) :=
  let fix loop m := if m is S i then f (loop i) else x in loop n.
Arguments iter !n f x /.

Lemma iterSr n f x : iter (S n) f x = iter n f (f x).
Proof.
(* unfolds [iter], does not reduce the fixpoint, and does not refold [iter]: *)
elim: n => //=.
Restart.
(* unfolds [iter], reduces the fixpoint, and does not refold [iter]: *)
elim: n => //. cbn.
Restart.
(* unfolds [iter], reduces the fixpoint, and refolds [iter]: *)
elim: n => //. rewrite /=.
Abort.

End Iteration.
```
ping @gares

**Comment by pi8027** [comment]
In the case of `move=> /=`, the `simpl` tactic is invoked by `tacred_simpl`: https://github.com/coq/coq/blob/155688103c43f578a8aef464bf0cb9a76acd269e/plugins/ssr/ssrequality.ml#L45-L49 and the `SimplIsCbn` flag is handled by `reduction_of_red_expr_val`: https://github.com/coq/coq/blob/155688103c43f578a8aef464bf0cb9a76acd269e/tactics/redexpr.ml#L254-L257

In the case of `rewrite /=`, the `simpl` tactic is invoked by `simplintac`: https://github.com/coq/coq/blob/155688103c43f578a8aef464bf0cb9a76acd269e/plugins/ssr/ssrequality.ml#L231 and it seems to ignore the `SimplIsCbn` flag (given that `Tacred.simpl` is called by `reduction_of_red_expr_val`): https://github.com/coq/coq/blob/155688103c43f578a8aef464bf0cb9a76acd269e/pretyping/tacred.ml#L1080-L1085

So, the issue seems twofold:
- `rewrite /=` ignores the `SimplIsCbn` flag, and
- `move=> /=` calls the `cbn` tactic if `SimplIsCbn` is set, but with a different set of flags?

Apparently, we can also conclude that `cbn` cannot refold `iter` in the second example anyways.

**Comment by herbelin** [comment]
> Apparently, we can also conclude that cbn cannot refold iter in the second example anyways. I guess I should open another issue for this point.

Indeed, `cbn` is not as powerful as `simpl` for refolding fixpoints which do not derive from a `Fixpoint` definition, as is the case in:
```coq
Definition iter n f x := let fix loop m := if m is S i then f (loop i) else x in loop n.
```
In particular `simpl` knows how to track arguments up to permutation but not `cbn`.

**Comment by pi8027** [comment]
@herbelin Do you think #18601 completely fixes the refolding issue in `cbn`? Then, we can probably try to replace `simpl` with `cbn` in MathComp. (Well, we still have to let `rewrite /=` respect the `SimplIsCbn` flag.) cc: @gares 

I think I will have time to try that only after the ITP deadline.

**Comment by herbelin** [comment]
@pi8027, sorry for the delay, I realize only now that you posted this comment.

> Do you think #18601 completely fixes the refolding issue in cbn

What #18601 does is only to address an issue with mutual fix. Regarding non-standard global fixpoints (e.g. of the form `Definition f ctx := (fix ...) args` where `args` is non empty), `cbn` refolding is still weaker than `simpl` refolding.

Regarding `cbn` vs `simpl`, there used to be a time where the consensus was to prefer going in the direction of `cbn`. I'm not sure that this is so clear now (for instance, `cbn` is currently slower than `simpl`, though optimization could be tried, see #18619). I'm not an active-enough user of `cbn` to compare it well with `simpl` but they are different, and, from what I hear, it is not clear that one is always better than the other. See e.g. this [Zulip](https://coq.zulipchat.com/#narrow/stream/237658-MetaCoq/topic/Cbn.20vs.20simpl.20in.20MetaCoq) discussion.

Finally, regarding `cbn` vs `simpl` specially in MathComp, I would rely on the experience of @gares.
